fsspec/implementations/tests/test_zip.py

```diff
@@ -47,3 +49,100 @@ def test_pickle():
         fs = fsspec.get_filesystem_class("zip")(fo=z)
         fs2 = pickle.loads(pickle.dumps(fs))
         assert fs2.cat("b") == b"hello"
+
+
+def test_all_dirnames():
+    with tempzip() as z:
+        fs = fsspec.get_filesystem_class("zip")(fo=z)
+
+        # fx are files, dx are a directories
+        assert fs._all_dirnames([]) == set()
+        assert fs._all_dirnames(["f1"]) == set()
+        assert fs._all_dirnames(["f1", "f2"]) == set()
+        assert fs._all_dirnames(["f1", "f2", "d1/f1"]) == {"d1"}
+        assert fs._all_dirnames(["f1", "d1/f1", "d1/f2"]) == {"d1"}
+        assert fs._all_dirnames(["f1", "d1/f1", "d2/f1"]) == {"d1", "d2"}
+        assert fs._all_dirnames(["d1/d1/d1/f1"]) == {"d1", "d1/d1", "d1/d1/d1"}
+
+
+def test_ls():
+    with tempzip(data) as z:
+        lhs = fsspec.get_filesystem_class("zip")(fo=z)
+
+        assert lhs.ls("") == ["a", "b", "deeply/"]
+        assert lhs.ls("/") == lhs.ls("")
```

Comment:
(OK, but the handling of paths in s3fs has been a constant headache, since it is backed by a simulated filesystem, and may not be the best example to follow!)